service/frontend/workflow_handler.go

```diff
@@ -884,6 +884,14 @@ func (wh *WorkflowHandler) PollWorkflowTaskQueue(ctx context.Context, request *w
 	}
 	namespaceID := namespaceEntry.ID()
 
+	// Copy WorkerVersionCapabilities.BuildId to BinaryChecksum if BinaryChecksum is missing (small
+	// optimization to save space in the poll request).
+	if request.WorkerVersionCapabilities != nil {
+		if len(request.WorkerVersionCapabilities.BuildId) > 0 && len(request.BinaryChecksum) == 0 {
+			request.BinaryChecksum = request.WorkerVersionCapabilities.BuildId
```

Comment:
Mutating the user's request looks off to me. I'd prefer we parse the generated request struct into a derived type upfront that meets all our invariants and is more ergonomic to work with. We already do this with TaskSerializer

Reply:
This request gets threaded all the way through to history so that's not really practical.

Mutating the request is not the cleanest, but this particular one is idempotent so it won't cause problems with retries. We could make a copy instead

Reply:
(To be clear, the request struct is fine as it is, there's nothing really un-ergonomic about it, and there are no new invariants: BinaryChecksum is not required. We're just doing a transport optimization)

Reply:
> This request gets threaded all the way through to history so that's not really practical.

I see, ok then sgtm